cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,688 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "compact_protocol_reader.hpp"
+#include "io/parquet/parquet.hpp"
+#include "reader_impl_helpers.hpp"
+
+#include <cudf/ast/detail/expression_transformer.hpp>
+#include <cudf/ast/detail/operators.hpp>
+#include <cudf/ast/expressions.hpp>
+#include <cudf/detail/cuco_helpers.hpp>
+#include <cudf/detail/transform.hpp>
+#include <cudf/hashing/detail/xxhash_64.cuh>
+#include <cudf/logger.hpp>
+#include <cudf/utilities/span.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_checks.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuco/bloom_filter_policies.cuh>
+#include <cuco/bloom_filter_ref.cuh>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/tabulate.h>
+
+#include <future>
+#include <numeric>
+#include <optional>
+
+namespace cudf::io::parquet::detail {
+namespace {
+
+/**
+ * @brief Converts bloom filter membership results (for each column chunk) to a device column.
+ *
+ */
+struct bloom_filter_caster {
+  cudf::device_span<cudf::device_span<cuda::std::byte> const> bloom_filter_spans;
+  host_span<Type const> parquet_types;
+  size_t total_row_groups;
+  size_t num_equality_columns;
+
+  enum class is_int96_timestamp : bool { YES, NO };
+
+  template <typename T, is_int96_timestamp IS_INT96_TIMESTAMP = is_int96_timestamp::NO>
+  std::unique_ptr<cudf::column> query_bloom_filter(cudf::size_type equality_col_idx,
+                                                   cudf::data_type dtype,
+                                                   ast::literal const* const literal,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::device_async_resource_ref mr) const
+  {
+    using key_type    = T;
+    using policy_type = cuco::arrow_filter_policy<key_type, cudf::hashing::detail::XXHash_64>;
+    using word_type   = typename policy_type::word_type;
+
+    // List, Struct, Dictionary types are not supported
```

Comment:
Are these types supported by other readers/writers? I would love to know if there is a specification for hashing compound types somewhere.

Reply:
Seems like nested types aren't supported by arrow either. Also reminds me to disable `boolean` cols for bloom filters as well.

https://github.com/apache/arrow/blob/32fcd184da91e5d9bc9098baeef4f368632fc1f1/cpp/src/parquet/bloom_filter_test.cc#L382C1-L384C83